### ISSUE TITLE
CustomAnalyzer Problem

### ISSUE THREAD

**Issue by gurkavcu** [jfp/play-search#10]
I have written a CustomAnalyzer to my play application. It locates in my application utils package .

package utils;
public class CustomAnalyzer extends Analyzer {
}

Then i configure my application.conf file like you described in the document :
play.search.analyser=utils.CustomAnalyzer

This way play couldnt find my CustomAnalyzer class and i got this exception:
java.lang.ClassNotFoundException: utils.CustomAnalyzer

If i create a jar library that contains CustomAnalyzer and put it in the application lib folder it is working. But i dont want to separate it from my application code.

In your plugin's Search class:

 public static Analyzer getAnalyser() {
        try {
            Class clazz = Class.forName(ANALYSER_CLASS);
            return (Analyzer) clazz.newInstance();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

"Class.forName(ANALYSER_CLASS) " cant locate main play application's class files. 

*\* I'am using play 1.1 and search-head. My play application can not  find your module unless i put your plugin libs to my application libs.

**Comment by jfp** [comment]
Yep, This is because of the class loader hierarchy in play, which currently is:
system loader -> play classpath -> your application's path

When trying to resolve a class, the application's path is first used, then play's jars, then system. Classes in your application are reloadable, jars are not. Reloading involves potential resource consumption, so it's better to have it only manage your app.

Inside the search plugin, you could try to :
- move the sources from the src folder to the app folder
- remove the lib/search.jar plugin

This way, it would have a chance to classload your analyser class. Another solution like you stated would be to put your analyser class inside a jar in your application or in the search module.

Regards,

**Comment by jmrenouard** [comment]
The code to patch is here.
Search.java
 public static Analyzer getAnalyser() {
        try {
            Class clazz = Class.forName(ANALYSER_CLASS);
            return (Analyzer) clazz.newInstance();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

Using Loading features in Play "should" solve this problem.
public static Analyzer getAnalyser() {
        try {
            Class clazz = ApplicationClassloader.getSystemClassLoader().loadClass(ANALYSER_CLASS);
           //OR
           Class clazz = Play.classes.getApplicationClass(ANALYSER_CLASS).javaClass;
    return (Analyzer) clazz.newInstance();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

Have a nice day.
JMR

**Comment by tomnatt** [comment]
Is this issues likely to be resolved? Although the workarounds here fix things it's a pain having to recompile the custom analyzers separately.

**Comment by jfp** [comment]
Hi,

Please try 16d7a4e3f37414e1cb2f7332fd7bb86af75ec0f5, it's in the head. If nobody has reloading issues / related problems, we'll keep it like that.

**Comment by tomnatt** [comment]
Hi

Apologies for the slow response. Patch is working well for us - thank you for fixing it.

**Comment by jmrenouard** [comment]
ok for me too.

**Comment by clacote** [comment]
Hi everyone. I'm also facing the same issue, and would like to use a custom analyzer as soon as possible (website already in production, but with search issues...).
Any plan for a release of this fixed version?

**Comment by tomnatt** [comment]
Fluor, if you use search head instead of an earlier version you will get the patches which enables the feature you need.

**Comment by clacote** [comment]
Sorry I didn't respond here : I did use "head" version, I my problem is was solved.
